iree/compiler/Dialect/Flow/Transforms/ConvertLinalgMatmulToMmt4D.cpp

```diff
@@ -98,45 +225,47 @@ class LinalgMatmulOpToLinalgMmt4DOpPattern
 
   LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
                                 PatternRewriter &rewriter) const override {
-    auto loc = matmulOp.getLoc();
+    Location loc = matmulOp.getLoc();
 
     Value lhs = matmulOp.getInputOperand(0)->get();
     Value rhs = matmulOp.getInputOperand(1)->get();
-    Value dst = matmulOp.getOutputOperand(0)->get();
-
-    RankedTensorType lhsType = lhs.getType().dyn_cast<RankedTensorType>();
-    RankedTensorType rhsType = rhs.getType().dyn_cast<RankedTensorType>();
-
-    if (!lhsType || !rhsType || !lhsType.hasStaticShape() ||
-        !rhsType.hasStaticShape()) {
+    Value acc = matmulOp.getOutputOperand(0)->get();
+
+    // This transformation supports any mixing of static and dynamic dimensions,
+    // with one exception: the dynamic-ness of each dimension of the accumulator
+    // must match the dynamic-ness of the corresponding lhs/rhs dimension.
+    // This limitation is not inherent to this transformation's code, it's just
+    // here to avoid a current linalg folding limitation: at the moment,
+    // removing this gives the following error in e2e matmul tests,
+    //   "error: failed to legalize operation 'tensor.cast' that was explicitly
+    //   marked illegal"
+    // apparently due to some missing folding of tensor.cast op into reshapes.
```

Comment:
This would be good to fix. Its a matter of just folding the tensor.cast away, so it should be a really simple one.

Reply:
ack, I will try to do it (and will then update/simplify this file).